hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ReplicationManager.java

```diff
@@ -512,25 +523,60 @@ private void handleUnderReplicatedContainer(final ContainerInfo container,
       if (source.size() > 0) {
         final int replicationFactor = container
             .getReplicationFactor().getNumber();
-        final int delta = replicationFactor - getReplicaCount(id, replicas);
+        // Want to check if the container is mis-replicated after considering
+        // inflight add and delete.
+        // Create a new list from source (healthy replicas minus pending delete)
+        List<DatanodeDetails> targetReplicas = new ArrayList<>(source);
+        // Then add any pending additions
+        targetReplicas.addAll(replicationInFlight);
+
+        int delta = replicationFactor - getReplicaCount(id, replicas);
+        final ContainerPlacementStatus placementStatus =
+            containerPlacement.validateContainerPlacement(
+                targetReplicas, replicationFactor);
+        final int misRepDelta = placementStatus.additionalReplicaRequired();
```

Comment:
This method in reality gives back the additional rack count required, not the additional replicas required.
This is why we need to check against the replication factor and the replica count difference also, while we provide the replication factor to the validation method, so it would be able to return the number of replicas really required.

Let see the different cases:
 - non-topology aware placement policies: if replica count == replication factor we are good, otherwise the required additional replica count is replication factor - replica count, and we need that many extra replicas later from chooseDataNodes.
 - a topology aware placement policy with replication factor of 1 in case it is under replicated we already miss the block
 - a topology aware placement policy with replication factor of 2 in case it is under replicated either we miss the block, or we have it in one rack, and chooseDatanodes will give us the required one in a different rack.
 - topology aware placement policies with replication factor of 3 (or more):
   - when we have 1 healthy replica, we need 2 additional replicas anyways, it is not really interesting in which rack.
   - when we have 2 healthy replica, we need 1 additional replica from either one of the racks in which we already have a replica or in an other rack instead if we have the two replica in the same rack, this the policy already handles as it uses the excluded node list to anchor to proper racks in chooseDatanodes().
   - when we have 3 healthy replica, but all three is in the same rack, then additionalReplicaRequired should yield to 1 still, and chooseDataNode should select a new set of DataNodes correctly, consisting of the nodes already have the replica, and one additional in a different rack. Which makes the container over replicated, so one excess replica will be removed later, after replication finished.

So if the additionalReplicaRequired works as the name suggests, we should not need to check replication factor and replica count here, but just rely on the value given back by the method.

Reply:
The PlacementStatus.additionalReplicaRequired() method is intended to state how many replicas are need to meet the placement policy and not the replication factor. In other places (eg Recon) it might be useful to be able to get the "additional racks required" separately from under replication.

Reply:
Where we have a difference is two cases with replication factor of 3:
 - we have only 1 replica, we need two replica in one rack (delta = 2, misRepDelta = 1)
 - we have 2 replica in different racks, we need one replica and zero rack (delta = 1, misRepDelta = 0)

I see this as something that still implicitly brings in rack awareness into ReplicationManager via an arbitrary int value, that comes from a method which has a name, that - for me at least - does not suggest this subtle but fundamental difference at all, that is why I started to check into this and suggested to get rid of it. I don't see the "it might be good to report this for Recon" as a good reason to increase the complexity, if we want this in Recon, we can add a new status implementation with this value and its accessor added for rack aware policies and we can grab the value from that, if it will be needed.
For the same reason (implicit rack awareness) I am unhappy about the check in lines from L562, and suggested a way there if we can achieve it.

I understand the statement "a container has a valid placement", as "we have the replication factor amount of replicas placed as we want them to be placed". And my suggestion reflects this understanding. Isn't this the good way to look at it? If not, why, what is the gain if we look it differently?

Reply:
> I understand the statement "a container has a valid placement", as "we have the replication factor amount of replicas placed as we want them to be placed"

Placement and replication are two different things, and I think it is good to keep them that way.

If you consider HDFS fsck - if a container is under replicated, it will report that. If it is mis-replicated it will report that. If it is both, it reports them both separately, along with their "missing count" for each. In HDFS for racks, there can only ever be 1 additional rack needed. But with upgrade domain, there can be possibly two additional domains needed.

> I see this as something that still implicitly brings in rack awareness into ReplicationManager via an arbitrary int value

It brings mis-replication into the RM, not rack awareness, and this is OK. A container needs to meet several goals to be properly replicated - the correct number of healthy replicas and the correct placement and this is what the replication manager checks.

In Recon, we want to be able to give an indicator about mis-replication and under replication separately in a similar way and ideally with Recon we just load the exact same placement policy as the cluster is using and call the same methods.

>  I am unhappy about the check in lines from L562, and suggested a way there if we can achieve it.

This check is still useful. If we combined the two rules as you are suggesting so "addtitionalReplicaRequired" return a single number representing delta and misRep, then in the case where you have delta=1, misRep=1 (ie two replicas on one rack), then if the newly selected node, due to fallback, gives us 3 replicas on one rack. This is OK, but the next time this container is checked, we have 3 replicas on one rack, and the logic will return 1, telling us to create another. We need to be sure this improves the mis-replication before creating a 4th replica, as that will make the container over-replicated.

Of course, we can check:

 * If we test 2 replicas with 1 rack, the logic will return 1.
 * If we test 3 replicas with 1 rack, the logic will return 1.
 * If we test 4 replicas with 1 rack, the logic will return 1.

So if we push the under and mis-replication numbers together, we cannot really see what is going on, and we will need to test the replication count too I think anyway.

I can give these methods new names or improve the java docs a bit more to make it clear that placement != replication and the number returned from the ContainerPlacementStatus object does not reflect replicas. I've made a small change to the doc now and will push it up:

```
  /**
   * Returns a boolean indicating if the container replicas meet the desired
   * placement policy. That is, they are placed on a sufficient number of
   * racks, or node groups etc. This does not check if the container is over
   * or under replicated, as it is possible for a container to have enough
   * replicas and still not meet the placement rules.
   * @return True if the containers meet the policy. False otherwise.
   */
  boolean isPolicySatisfied();

  /**
   * Returns an String describing why a container does not meet the placement
   * policy.
   * @return String indicating the reason the policy is not satisfied or null if
   *         the policy is satisfied.
   */
  String misReplicatedReason();

  /**
   * If the container does not meet the placement policy, return an integer
   * indicating how many additional replicas are required so the container
   * meets the placement policy. Otherwise return zero.
   * Note the count returned are the number of replicas needed to meet the
   * placement policy. The container may need additional replicas if it is
   * under replicated. The container could also have sufficient replicas but
   * require more to make it meet the policy, as the existing replicas are not
   * placed correctly.
   * @return The number of additional replicas required, or zero.
   */
  int additionalReplicaRequired();
```

I could change the name of additionalReplicaRequired() to be something like replicasRequiredForPlacement(), if yo think it makes it more clear.

Reply:
Ah ok, finally I understood :) Sorry for being a bit hard headed on this one, but it seems I haven't thought of the ambiguity in the mentioned case which is needed to be handled unless we will run into an infinite loop with the current fallback logic. I still feel that this we might revise when/if we touch fallback logic, but unsure :)

I would be glad if the name can be changed though, I don't have a better idea then the suggested one though... maybe misReplicationCount()?

Also in the javadoc, in the last sentence before @return do you mean "placed incorrectly"?

Reply:
Do you mean this line:

> The container could also have sufficient replicas but require more to make it meet the policy, as the existing replicas are **not** placed correctly.

Did you miss the "not" in there? :)

I will change this method to be called misReplicationCount() which hopefully makes it more clear.

Reply:
oh... I did, it was hiding behind a scrollbar I haven't scrolled here in the comment stream :)

Looks good this way ;)